src/qibolab/instruments/qm.py

```diff
@@ -0,0 +1,598 @@
+import collections
+from types import SimpleNamespace
+
+import numpy as np
+from qibo.config import log, raise_error
+from qm.QuantumMachinesManager import QuantumMachinesManager
+
+from qibolab.instruments.abstract import AbstractInstrument
+from qibolab.result import ExecutionResult
+
+
+class QMOPX(AbstractInstrument):
+    """Instrument object for controlling Quantum Machines (QM) OPX controllers.
+
+    Playing pulses on QM controllers requires a ``config`` dictionary and a program
+    written in QUA language. The ``config`` file is generated in parts in the following places
+    in the ``register_*`` methods. The controllers, elements and pulses are all
+    registered after a pulse sequence is given, so that the config contains only
+    elements related to the participating qubits.
+    The QUA program for executing an arbitrary qibolab ``PulseSequence`` is written in
+    ``play`` and ``play_pulses`` and executed in ``execute_program``.
+
+    Args:
+        name (str): Name of the instrument instance.
+        address (str): IP address and port for connecting to the OPX instruments.
+
+    Attributes:
+        is_connected (bool): Boolean that shows whether instruments are connected.
+        manager (:class:`qm.QuantumMachinesManager.QuantumMachinesManager`): Manager object
+            used for controlling the QM OPXs.
+        config (dict): Configuration dictionary required for pulse execution on the OPXs.
+        time_of_flight (optional,int): Time of flight used for hardware signal integration.
+        smearing (optional,int): Smearing used for hardware signal integration.
+    """
+
+    def __init__(self, name, address):
+        # QuantumMachines manager is instantiated in ``platform.connect``
+        self.name = name
+        self.address = address
+        self.manager = None
+        self.is_connected = False
+
+        self.time_of_flight = 0
+        self.smearing = 0
+        # copied from qblox runcard, not used here yet
+        # hardware_avg: 1024
+        # sampling_rate: 1_000_000_000
+        # repetition_duration: 200_000
+        # minimum_delay_between_instructions: 4
+
+        # Default configuration for communicating with the ``QuantumMachinesManager``
+        self.config = {
+            "version": 1,
+            "controllers": {},
+            "elements": {},
+            "pulses": {},
+            "waveforms": {},
+            "digital_waveforms": {
+                "ON": {"samples": [(1, 0)]},
+            },
+            "integration_weights": {},
+            "mixers": {},
+        }
+
+    def connect(self):
+        host, port = self.address.split(":")
+        self.manager = QuantumMachinesManager(host, int(port))
+
+    def setup(self, qubits, time_of_flight=0, smearing=0):
+        self.time_of_flight = time_of_flight
+        self.smearing = smearing
+        # controllers are defined when registering pulses
+
+    def start(self):
+        # TODO: Start the OPX flux offsets?
+        pass
```

Comment:
If you are not doing anything, you should not need to define the module.

I would suggest to set something like `is_running`, similar to `is_connected`, to keep track of the status of the design.
Even better, I would replace `is_connected` with a `status` attribute, of type `Status`, and make `Status` an [enum](https://docs.python.org/3/library/enum.html#module-enum):
```py
from enum import Flag, auto

class Status(Flag):
    CONNECTED = auto()
    RUNNING = auto()
```